Lib/pathlib/_types.py

```diff
@@ -0,0 +1,56 @@
+"""
+Protocols for supporting classes in pathlib.
+"""
+from typing import Protocol, runtime_checkable
+
+
+@runtime_checkable
+class Parser(Protocol):
+    """Protocol for path parsers, which do low-level path manipulation.
+
+    Path parsers provide a subset of the os.path API, specifically those
+    functions needed to provide PurePathBase functionality. Each PurePathBase
+    subclass references its path parser via a 'parser' class attribute.
+    """
+
+    sep: str
+    def join(self, path: str, *paths: str) -> str: ...
+    def split(self, path: str) -> tuple[str, str]: ...
+    def splitdrive(self, path: str) -> tuple[str, str]: ...
+    def splitext(self, path: str) -> tuple[str, str]: ...
+    def normcase(self, path: str) -> str: ...
+    def isabs(self, path: str) -> bool: ...
+
+
+@runtime_checkable
+class DirEntry(Protocol):
```

Comment:
@zooba continuing our discussion from #127377 -- this is the protocol that must be implemented by objects generated from `PathBase.scandir()`.

I think it's a neat and simple protocol for storing possibly-cached information about directory children. It's not too difficult to implement I don't think.

Users who want to get started quickly would override only `PathBase.iterdir()`.

Users who want more performance can additionally implement `PathBase.scandir()` to generate custom directory entry objects with cached information, as long as they implement the `pathlib._types.DirEntry` protocol.

I'm writing all this because I *think* we were using "DirEntry" in different ways in the other discussion. But if you already understood all this, and it's simply my failure to comprehend your feedback, then I do apologise :sweat_smile:

Reply:
Yeah, I see what you're trying to do, I just think it's optimising at the wrong place.

Right now, you're designing for the case where the best way to implement `glob` is "like a Windows-style file system", and providing the hooks for other types of systems to fit that model (specifically, where the most efficient way to search a directory is non-recursive iteration that includes item type information).

My contention is that that's a fairly rare case, and not worth the additional optimisation.

For cases more like a *nix-style file system, the hooks don't need to include type information, since obtaining it separately is just as efficient. All the `scandir`-like hooks do is force someone to wrap up two operations into one, when the two would've been just fine.

For cases more like any online file storage, it's unlikely that the most efficient approach is the same as a Windows file-system. I would expect some kind of query to exist to pre-filter before returning results, and probably recursively. So the `scandir` like hook isn't helpful, because you can implement `glob` far more efficiently (and probably `copy` as well).

I don't think the scenarios that benefit from `scandir` are generic enough to be worth making it the extension point. Extending at `glob` for performance or `iterdir` for convenience is sufficient. We don't need to grow the API just for the sake of the case in the middle.

Reply:
This is a minor point, but the docs for [`os.DirEntry.is_dir()`](https://docs.python.org/3/library/os.html#os.DirEntry.is_dir) say that it doesn't require a system call in most cases _even on POSIX_. Am I missing something here? Surely using `os.scandir()` isn't a Windows-only optimisation?

Reply:
Looks like it uses [`readdir`](https://www.man7.org/linux/man-pages/man3/readdir.3.html) on POSIX, which also returns some stat information.

My point was more that you've got no reason to assume the effect will extend to S3 buckets or other hierarchical storage, or that a `scandir`-like hook is easier than overriding `glob` would be. This hook is over-specialising, where there are already two choices for specialising these operations.

Reply:
I have _some_ reason to assume that - the [Artifactory API](https://jfrog.com/help/r/jfrog-rest-apis/get-folder-info) tells me whether each child is a folder, and my toy implementation of `ZipPath` also has this property.

Reply:
I think I see your point though, thanks for being patient with me. I'll give it some thought.

Reply:
Looks like GitHub's is similar, but both match the `walk` interface better than `scandir`. Perhaps making `walk` the root of the implementation makes more sense?